gtars/src/uniwig/mod.rs

```diff
@@ -1,16 +1,722 @@
+            file.write_all(b"\n").unwrap();
+            position+=1;
+        }
+
+    }
+
+
+}
+
+fn read_chromosome_sizes(chrom_size_path: &str) -> Result<std::collections::HashMap<String, u32>, Box<dyn Error>> {
+    let chrom_size_file = File::open(Path::new(chrom_size_path))?;
+    let mut chrom_sizes = std::collections::HashMap::new();
+    let reader = BufReader::new(chrom_size_file);
+
+    for line in reader.lines() {
+        let line = line?; // Propagate the potential error
+        let mut iter = line.split('\t');
+        let chrom_name = iter.next().unwrap().to_owned();
+        let _ = iter.next().unwrap();
+        let size_str = iter.next().unwrap(); // we really want the 3rd column which is the end column.
+        let size = size_str.parse::<u32>()?;
+
+        chrom_sizes.insert(chrom_name, size);
+    }
+
+    Ok(chrom_sizes)
+}
+
+#[allow(unused_variables)]
+pub fn smooth_fixed_start_end_wiggle(starts_vector: &Vec<i32>, chrom_size: i32, smoothsize: i32, stepsize:i32) -> (Vec<u32>, Vec<i32>) {
+    // This function is a more direct port of smoothFixedStartEndBW from uniwig written in CPP
+    // It allows the user to accumulate reads of either starts or ends
+    // Counts occur between a start coordinate (cutSite) and an end site (endSite) where the endsite is determined based on
+    // the level of smoothing.
+    // counts are reported over a stepsize (with a default of stepsize = 1)
+    // Unlike the original function, it does not write to disk in chunks. it simply returns a vector of accumulated reads.
+
+
+
+    println!("BEGIN smooth_Fixed_Start_End_Wiggle");
+
+    let vin_iter = starts_vector.iter();
+
+    let mut v_coordinate_positions: Vec<i32> = Vec::new(); // these are the final coordinates after any adjustments
+    let mut v_coord_counts: Vec<u32> = Vec::new(); // u8 stores 0:255 This may be insufficient. u16 max is 65535
+
+    let mut coordinate_position = 1;
+
+    let mut count:u32 = 0;
+
+    let mut coordinate_value = 0;
+    let mut prev_coordinate_value = 0;
+
+
+    let mut adjusted_start_site =0;
+    let mut current_end_site = 0;
+
+    let mut collected_end_sites: Vec<i32> = Vec::new();
+
+    //println!("DEBUG: START SITE BEFORE ADJUSTMENT -> {}",starts_vector[0].clone());
+
+    adjusted_start_site = starts_vector[0].clone(); // get first coordinate position
+    adjusted_start_site = adjusted_start_site - smoothsize; // adjust based on smoothing
+    //println!("DEBUG: START SITE AFTER ADJUSTMENT -> {}",adjusted_start_site.clone());
+
+    //Check endsite generation
+    current_end_site = adjusted_start_site + 1 + smoothsize*2;
+
+    //println!("DEBUG: INITIAL ENDSITE -> {}", current_end_site.clone());
+
+    if adjusted_start_site < 1{
+        adjusted_start_site = 1;
+    }
+
+    //println!("DEBUG: SKIPPING UNTIL COORDINATE_POSITION < ADJUSTEDSTARTSITE -> {}  {}", coordinate_position.clone(), adjusted_start_site.clone());
+    while coordinate_position < adjusted_start_site{
```

Comment:
Any condition to stop if something goes horribly wrong? haha

Reply:
Glancing at this, we could probably break this if we gave it a _negative_ step size, so we could ensure that is not possible.